From: Jed <lejosnej@ainfosec.com>
Date: Fri, 23 Jan 2015 16:54:42 -0500
Subject: [PATCH] qemu: add a patch for OpenXT audio policy bits, with just the audio recording flag for now

################################################################################
SHORT DESCRIPTION: 
################################################################################
Add a patch for OpenXT audio policy bits, with just the audio recording flag
for now.

################################################################################
LONG DESCRIPTION: 
################################################################################
TODO

################################################################################
CHANGELOG 
################################################################################
Ported to QEMU 2.6.2: Ross Philipson, philipsonr@ainfosec.com, 10/13/2016

################################################################################
REMOVAL 
################################################################################

################################################################################
UPSTREAM PLAN 
################################################################################

################################################################################
INTERNAL DEPENDENCIES 
################################################################################

################################################################################
PATCHES
################################################################################
Index: qemu-2.6.2/hw/audio/ac97.c
===================================================================
--- qemu-2.6.2.orig/hw/audio/ac97.c
+++ qemu-2.6.2/hw/audio/ac97.c
@@ -24,6 +24,8 @@
 #include "hw/pci/pci.h"
 #include "sysemu/dma.h"
 
+extern int disable_audio_recording;
+
 enum {
     AC97_Reset                     = 0x00,
     AC97_Master_Volume_Mute        = 0x02,
@@ -1035,6 +1037,13 @@ static int read_audio (AC97LinkState *s,
     int to_copy = 0;
     SWVoiceIn *voice = (r - s->bm_regs) == MC_INDEX ? s->voice_mc : s->voice_pi;
 
+    /* Disable audio recording. */
+    if (disable_audio_recording)
+    {
+        *stop = 1;
+        return 0;
+    }
+
     temp = audio_MIN (temp, max);
 
     if (!temp) {
Index: qemu-2.6.2/hw/audio/hda-codec.c
===================================================================
--- qemu-2.6.2.orig/hw/audio/hda-codec.c
+++ qemu-2.6.2/hw/audio/hda-codec.c
@@ -26,6 +26,8 @@
 
 /* -------------------------------------------------------------------------- */
 
+extern int disable_audio_recording;
+
 typedef struct desc_param {
     uint32_t id;
     uint32_t val;
@@ -193,10 +195,13 @@ static void hda_audio_input_cb(void *opa
                 break;
             }
         }
-        rc = hda_codec_xfer(&st->state->hda, st->stream, false,
-                            st->buf, sizeof(st->buf));
-        if (!rc) {
-            break;
+        /* Only transfer input if allowed by the policy */
+        if (!disable_audio_recording) {
+            rc = hda_codec_xfer(&st->state->hda, st->stream, false,
+                                st->buf, sizeof(st->buf));
+            if (!rc) {
+                break;
+            }
         }
         st->bpos = 0;
     }
Index: qemu-2.6.2/qemu-options.hx
===================================================================
--- qemu-2.6.2.orig/qemu-options.hx
+++ qemu-2.6.2/qemu-options.hx
@@ -359,6 +359,14 @@ modprobe i810_audio clocking=48000
 @end example
 ETEXI
 
+DEF("disable-audio-rec", 0, QEMU_OPTION_disable_audio_rec,
+    "-disable-audio-rec Disable audio recording\n", QEMU_ARCH_ALL)
+STEXI
+@item -disable-audio-rec
+@findex -disable-audio-rec
+Disable audio recording in the guest.
+ETEXI
+
 DEF("balloon", HAS_ARG, QEMU_OPTION_balloon,
     "-balloon none   disable balloon device\n"
     "-balloon virtio[,addr=str]\n"
Index: qemu-2.6.2/vl.c
===================================================================
--- qemu-2.6.2.orig/vl.c
+++ qemu-2.6.2/vl.c
@@ -182,6 +182,7 @@ bool boot_strict;
 uint8_t *boot_splash_filedata;
 size_t boot_splash_filedata_size;
 uint8_t qemu_extra_params_fw[2];
+int disable_audio_recording = 0;
 
 int icount_align_option;
 
@@ -3341,6 +3342,9 @@ int main(int argc, char **argv, char **e
             case QEMU_OPTION_soundhw:
                 select_soundhw (optarg);
                 break;
+            case QEMU_OPTION_disable_audio_rec:
+                disable_audio_recording = 1;
+                break;
             case QEMU_OPTION_h:
                 help(0);
                 break;
